lib/elixir_analyzer/exercise_test/assert_call/compiler.ex

```diff
@@ -373,4 +383,48 @@ defmodule ElixirAnalyzer.ExerciseTest.AssertCall.Compiler do
         else: %{acc | modules_in_scope: Map.put(acc.modules_in_scope, alias, full_path)}
     end)
   end
+
+  # track all called functions
+  def track_all_functions(
+        %{function_call_tree: tree, in_module: module, in_function_def: name} = acc,
+        {_, _, _} = function
+      )
+      when not is_nil(name) do
+    called =
+      case function do
+        {:., _, [{:__MODULE__, _, _}, fn_name]} -> {module, fn_name}
+        {:., _, [{:__aliases__, _, fn_module}, fn_name]} -> {fn_module, fn_name}
+        {fn_name, _, _} -> {module, fn_name}
+      end
+
+    %{acc | function_call_tree: Map.update(tree, {module, name}, [called], &[called | &1])}
+  end
+
+  def track_all_functions(acc, _node), do: acc
+
+  # Check if a function was called through helper functions
+  def indirect_call?(%{called_fn: called_fn, calling_fn: calling_fn, function_call_tree: tree}) do
+    #    IO.inspect({called_fn, calling_fn, tree})
+
+    cond do
+      is_nil(tree[calling_fn]) ->
+        false
+
+      called_fn in tree[calling_fn] ->
+        true
+
+      true ->
+        tree[calling_fn]
+        # Delete function itself once because it is always added from the `def` node
+        |> List.delete(calling_fn)
+        |> Enum.any?(
+          &indirect_call?(%{
+            called_fn: called_fn,
+            calling_fn: &1,
+            # Remove tree branch already explored in case of cyclical calls
+            function_call_tree: Map.delete(tree, calling_fn)
+          })
+        )
+    end
```

Comment:
It took me a while to understand those branches. Maybe some comments would help? Also, do those comments make sense, did I understand correctly?
```suggestion
    cond do
      # calling_fn was never even called
      is_nil(tree[calling_fn]) ->
        false
	  
      # called_fn was called directly by calling_fn
      called_fn in tree[calling_fn] ->
        true

      # called_fn wasn't called directly by calling_fn, look for indirect calls
      true ->
        tree[calling_fn]
        # Delete function itself once because it is always added from the `def` node
        |> List.delete(calling_fn)
        |> Enum.any?(
          &indirect_call?(%{
            called_fn: called_fn,
            calling_fn: &1,
            # Remove tree branch already explored in case of cyclical calls
            function_call_tree: Map.delete(tree, calling_fn)
          })
        )
    end
```

Reply:
You got the idea, but because `indirect_call?` calls itself recursively, it's a little bit more subtle I guess? I will add comments.